Mll command line 101 updates (#103)
arcus/education_modules: bash_scripting_101/bash_scripting_101.md

<!--

author:   Nicole Feldman and Elizabeth Drellich
email:    feldmanna@chop.edu drelliche@chop.edu
version:  1.1.0
module_template_version: 2.0.0
language: en
narrator: UK English Female

> title: Bash 101

Reviewer: good call!

## The bash shell

Your computer has an **O**perating **S**ystem (OS), most commonly either Windows, Mac, or Linux. The core for the OS itself is called the kernel; it's the heart of how everything works in your computer, but you can't actually interface with it directly as a user. Programs that are designed to work with your OS can communicate with the kernel (Word, RStudio, Chrome, etc.), but if you want to do something outside of one of those programs you can't connect to the kernel directly yourself. That's where a shell comes in.

The shell is like a layer outside the kernel that you as a user can communicate with. You can type commands in the shell, and it will execute them for you in the kernel. Back when computers were newer and there were few ready programs available, the shell was the primary way to get stuff done on a computer.

<div class = "learnmore">
To learn more, check out this post for [an excellent breakdown of the difference between the shell and the kernel](https://www.geeksforgeeks.org/difference-between-shell-and-kernel/).
</div>
![Home directory of the shell in the default homebrew theme.](media/shell_home.png)

### What is it?

> A **C**ommand **L**ine **I**nterface or CLI allows a user to pass a series of text (commands) to a program in order to execute code. The execution of shell commands is known as shell scripting.

Reviewer: @pm0kjp
I restructured this section away, and think the new version is clearer. Would love your feedback!

[[X]] Shell scripting is a way for a user to interact with the core (or kernel) of your operating system.
[[X]] Command line interfaces are pre-installed on most operating systems.
[[ ]] Bash is expensive.
[[X]] Shell scripts can work with other programming languages, such as R and Python, in pipelines.
***
<div class = "answer">
Bash, or shell, scripting is a way to interface with your computer's operating system, and the necessary applications are probably already on your computer! You can also create pipelines that utilize bash and other programming languages. It's also 100% free!
</div>
***

## Navigating in bash

If you are new to bash scripting, it can take some adjustment to navigate by typing text. Once you get the hang of a few basic commands, this can be an easier way to deal with files and folders than dropdown menus and dragging icons from place to place. Even the language we use is a little different. In bash scripting you will hear folders refered to as **directories**.

To enter a command into the shell, type it in where the blinking cursor appears and press `Enter` or `return` on your keyboard.

### Navigation Commands

> Just like when you navigate any file system, it is important to know where you are, see what files and folders are in that location, and move to other locations.

Reviewer 2: ```suggestion
Just like when you navigate any file system, it is important to know where you are, see what files and folders are in that location, and know how move to other locations.
```

**Where am I?**

> The command `pwd`, which stands for **p**resent **w**orking **d**irectory, will return the path through your directory system to your current location.

Reviewer 2: Consider a screenshot here to make this more concrete

Reviewer: added a code block, let me know what you think.

**What is here?**

> The bash shell doesn't show the icons of the files and directories in your present working directory. Instead you will need to ask for a list of the directory's contents with the command `ls`.

Reviewer 2: Consider a screenshot here to make this more clear

**How do I move around?**

The command `cd` lets you **c**hange **d**irectory. Unlike the previous two commands, `cd` requires you to provide additional information: the directory you want to move to. You can move to a location relative to your current location, or you can move anywhere directly if you know the path to get there.

> - Move to a directory relative to your present working directory:

Reviewer 2: ```suggestion
- Navigate to a directory relative to your present working directory:
```
"move" is ambiguous because we also "move" files.

  * If the directory `directory_name` is in your present working directory (i.e. it is one of the listed items when you enter `ls`) you can navigate into it with the command `cd directory_name`

>   * To move to the parent directory of your current location, use `cd ..`

Reviewer 2: ```suggestion
  * To navigate to the parent directory of your current location, use `cd ..`
```
"move" is ambiguous because we "move" files also.  Just a thought!

>   * Use slashes `/` to link directions. For example `cd ../..` up to the the parent directory of your parent directory.

Reviewer 2: ```suggestion
  * Use slashes `/` to link more than one navigation. For example `cd ../..` will navigate you up to the the parent directory (`..`) of your parent directory (`..`).
```

>   * It is possible to move up and then back down in your file hierarchy. For example `cd ../other_directory` would move you to the `other_directory` that shares the same parent directory as your current location.

Reviewer 2: ```suggestion
  * It is possible to navigate up and then back down in your file hierarchy. For example `cd ../other_directory` would navigate you to the `other_directory` that shares the same parent directory as your current location.
```

Reviewer 2: This is just a nice-to-have, but especially for this example, a visual map showing the up arrow to "parent directory" and the down arrow to "other directory" might be useful.

Reviewer: @pm0kjp 
Do you have a map in mind? I always find these maps more confusing that useful, so I don't have a good idea of what others would find useful.

> - Move to a directory using its file path:

Reviewer 2: ```suggestion
- Navigate to a directory using its file path:
```

  * To return to your home directory use `cd ~`

  * If you know how to get to the directory you want from your home directory, you can go directly there with `cd ~/fixed/path/to/directory`

> **Give it a try!** Use these commands to explore the directories on your computer. Try navigating around some files and folders that you are already familiar with to get comfortable with how they look through the bash shell.

Reviewer 2: ```suggestion
**Give it a try!** Use `cd`, `ls`, and `pwd` to explore the directories on your computer. Try navigating around some files and folders that you are already familiar with to get comfortable with how they look through the bash shell.
```

### Tab completion

> If it seems like there is an awful lot to make sure you are typing correctly  into your bash shell, you are right. Luckily you don't have to type everything out fully because you can use the `tab` key on your computer.

Reviewer 2: ```suggestion
If it seems like there is an awful lot to make sure you are typing correctly into your bash shell, you are right. Luckily you don't have to type everything out fully because you can use the `tab` key on your computer.
```

> Once you start typing the first few letters of a directory, press `tab`. The shell will automatically fill in as much as it can of what you are typing. For example if I have only two directories in my current location, named `my_directory` and `shared_directory`, I don't need to type out the full name of the directory I want to `cd` into. In fact since they start with different letters, `cd m` followed by the `tab` instead of the `enter` or `return` key will automatically be completed to the only possibility, `cd my_directory`.

Reviewer 2: ```suggestion
Once you start typing the first few letters of a directory, press `tab`. The shell will automatically fill in as much as it can of what you are typing in a process known as **tab completion**. For example, if you have only two directories in your current location, named `my_directory` and `shared_directory`, you don't need to type out the full name of the directory you want to `cd` into. In fact, since they start with different letters, `cd m` followed by the `tab` instead of the `enter` or `return` key will automatically be completed to the only possibility, `cd my_directory`.
```
Just trying to avoid first person singular, this is a minor stylistic preference I have.

  * You can link the three steps together with `cd ../../data`
</div>
***

## Editing Files

The bash shell can see and move files of all types, but it is most useful for interacting directly with plain text files.

A [**plain text**](https://en.wikipedia.org/wiki/Plain_text) file is a file that contains only text characters like the ones you can type directly using your keyboard. For example the words "plain text" at the beginning of this paragraph, which are both bolded and hyperlinked, are not being presented to you as plain text. You might have run into problems with files that aren't in plain text if, for example, you ever opened a file in a new program and lost some of the formatting like particular fonts in Word or conditional highlighting in Excel.

<div class = "learnmore">
Even though the words "plain text" in the previous paragraph have some non-text attributes, the are all recorded as plain text in the Markdown (`.md`) file of this lesson. If you open the [Markdown file](link/to/git/of/this/file), you will see only characters that can be typed by a keyboard:

```
[**plain text**](https://en.wikipedia.org/wiki/Plain_text)
```
</div>

> If you have been using a particular program to write and edit your files like Word, RStudio, or Jupyter notebook, you will need to continue to use that program to edit those files. If you expect your project to grow substantially in size or number of files, having as much of your work as possible in plain text files may make scaling up easier.

Reviewer 2: "If you have been using a particular program to write and edit your files like Word, RStudio, or Jupyter notebook, you will need to continue to use that program to edit those files."

Why?

Reviewer 2: I guess what I'm getting at is you can always change your tooling -- if I was using RStudio I can actually switch over and write in Atom, etc., but I think I get your point.  Special file types tend to lock you in...

Some common types of plain text files you might have seen before are:

| file ending | description |
|:---:| :---|
| `.txt` | text file |
| `.csv` | comma separated values |
| `.md` | markdown file |
| `.html` | html file |

<div class = "learnmore">

> The file endings like `.txt`, `.csv`, or `.doc` (Word), `.xslx` (Excel) are the part of the file name that lets your computer know what type of program to use to open it. The ending doesn't actually impact the contents of the file, which is why we were able to create files like `my_file` in the previous section with no file ending at all.

Reviewer 2: ```suggestion
The file endings (or **extensions**) like `.txt`, `.csv`, or `.doc` (Word), `.xslx` (Excel) are the part of the file name that lets your computer know what type of program to use to open it. The ending doesn't actually impact the contents of the file, which is why we were able to create files like `my_file` in the previous section with no file ending at all.
```

<div class = "warning">
You might have noticed that none of the file or directory names we used had spaces in them. Bash scripting gets quite a bit more complicated when you have spaces in names, so it is considered best practice to avoid them.
</div>

### Writing to a file directly

We have created 4 files, but so far all of them are empty. The most direct way to write text to a file is to use the `echo` and `>>` commands.

**`echo`**

The bash command `echo` returns any text that follows it. Give it a try in your command line interface!

**`>>`**

The bash command `>>` is used to write the output of a command to the end of a file. For example the following code will create a file called `my_sentences` and write the text `This is a sentence.` in that file.

```
echo "This is a sentence." >> my_sentences
```

If you then want to add an additional line to `my_sentences` you can use the same method.

```
echo "This is another sentence." >> my_sentences
```

Now the file `my_sentences` has two lines:

```
This is a sentence.
This is another sentence.
```

**Give it a try!** Add some text to the files you already created!

<div class = "warning">

> Make sure you use the two arrows in `>>`. Using only a single arrow will also write text to your file, but instead of appending it to the end of the file, it will replace the entire contents of the file with that new text.

Reviewer 2: ```suggestion
Make sure you use the two greater than signs (here, we'll call them "arrows") in `>>`. Using only a single arrow will also write text to your file, but instead of appending it to the end of the file, it will replace the entire contents of the file with that new text.
```

<div class = "help">

**Stuck in Vim?**

The default text editor on macOS and Linux is usually set to Vim, which is not famous for being intuitive. That is why we don't recommend it as a text editor. If you accidentally find yourself stuck in it, hit the `Esc` key, followed by `:`+`Q`+`!` (colon, lower-case 'q', exclamation mark), then hitting Return to return to the shell.
</div>

### Quiz: Editing files

1. You want to make a new folder inside `learning_bash` called `data` and put two files in it, `data_1` and `data_2`.

```
cd ~/learning_bash
______A_______
______B_______
touch data_1 data_2
```
What code should go in lines A and B?

Line A should be:

[[ `touch data`| (`mkdir data`) | `cd data` ]]

Line B should be:

[[ `touch data`| `mkdir data` | (`cd data`) ]]
***
<div class = "answer">
You must first make the directory `data` and then move into that directory with `cd data`.

The command `touch data` will create a file named `data`, not a directory that can contain files.
</div>
***

> 2. Now that you have created the files `data_1` you want to put some content into it. How would you make the file `data_1` contain the text `sample 1: 45`?

Reviewer 2: This isn't working for me, I get the error "Unexpected identifier 'sample'. Expected ';' after variable declaration."

Reviewer: @pm0kjp 
Are you getting this error in your CLI? or in entering the answer in Liascript? I haven't been able to reproduce it.

Reviewer: I never got exactly the same error message you got, but I changed the whole format of the question to multiple choice because it seems having input that includes `"` and also allowing for multiple correct answers is every so slightly too hard.

[[echo "sample 1: 45" >> data_1]]
<script>
let input = "@input".trim();
  input == "echo \"sample 1: 45\" >> data_1" || input == "echo sample 1: 45 >> data_1" || input == "echo \"sample 1: 45\" > data_1" || input == "echo sample 1: 45 > data_1";
</script>
***
<div class = "answer">
The command `echo "sample 1: 45" >> data_1` will add `sample 1: 45` as a new line at the end of the current contents of the file `data_1`.

Since there wasn't anything in `data_1` to begin with, you will get the same outcome if you use a single `>` to replace the entire file with `sample 1: 45`.

The quotes around the text can also be safely omitted in this case, but that isn't a good habit to get into since you may need them later as you progress in your bash scripting.
</div>
***

## Additional Resources

- [Brief Illustration of the Difference between Shell and Kernel](https://www.geeksforgeeks.org/difference-between-shell-and-kernel/)

- [Reinforce Your New Knowledge through this Learning the Shell Page](https://linuxcommand.org/lc3_learning_the_shell.php)

### The bash manual

> The bash manual is available to you whenever you are in a bash shell.  Entering `man command_name` reveals the documentation for the function `command_name`.  This can be useful if you think a command might be useful and want to learn more about it, or have forgotten how to use a command or what its options are.

Reviewer 2: ```suggestion
The bash manual is available to you whenever you are in a bash shell.  Entering `man command_name` reveals the documentation (`man` is short for "manual") for the function `command_name`.  This can be useful if you think a command might be useful and want to learn more about it, or have forgotten how to use a command or what its options are.
```

Also include info here about how to exit and advance in man pages

Reviewer: good call!

> - Search Bash in Cortana to launch WSL for the first time and click “Y” when asked to install Ubuntu for the first time. [Microsoft has their own internal documentation on WSL which features additional visual resources and FAQS](https://docs.microsoft.com/en-us/windows/wsl/).

Reviewer: The Windows Installation needs to be updated for users who do not have admin access.

> - Search Bash in Cortana to launch WSL for the first time and click “Y” when asked to install Ubuntu for the first time. [Microsoft has their own internal documentation on WSL which features additional visual resources and FAQS](https://docs.microsoft.com/en-us/windows/wsl/).

Reviewer: The Windows Installation needs to be updated for users who do not have admin access.

- Create a username and password when prompted and now you are ready to do some shell scripting! These user credentials have no bearing on your Windows user name and you will be automatically logged in for future sessions.

### Quiz: Accessing the bash shell

What is the name of the pre-installed program used to run shell scripts on a Mac?

[( )] Together
[( )] MSL (Mac Subsystem for Linux)
[(X)] Terminal
[( )] Ubuntu
***
<div class = "answer">

WSL is the Windows utility for running shell. Ubuntu is a well known Linux operating system that is often used as the default shell scripting environment.

</div>
***

> ## Setting up a User/Home Directory

Reviewer: I'm not sure if this section is supposed to walk me through doing something, or show me where to find something that already exists. If a user doesn't have CLI experience, how would they find out if they are in their home directory or navigate to it? 

Maybe this page could contain instructions on how to navigate using `cd`, as well as `ls` and `pwd`

> ## Setting up a User/Home Directory

Reviewer: I'm not sure if this section is supposed to walk me through doing something, or show me where to find something that already exists. If a user doesn't have CLI experience, how would they find out if they are in their home directory or navigate to it? 

Maybe this page could contain instructions on how to navigate using `cd`, as well as `ls` and `pwd`

> 

Reviewer: Can we give examples of these commands? This reads more like a reference list that goes at the end of the module, rather than a "here are these things for the first time" presentation.

> 

Reviewer: Can we give examples of these commands? This reads more like a reference list that goes at the end of the module, rather than a "here are these things for the first time" presentation.

  ![Head output of slightly less basic python script that organizes a picnic gathering menu amongst three friends.](media/head_example_small.png)

### Quiz: Bash commands

Select the three functions of the `cat` command?

[[X]] View
[[ ]] Categorize
[[X]] Concatenate
[[X]] Create New Version
***
<div class = "answer">

The `sort` command is used to categorize files in bash.

</div>
***

> ### Syntax

Reviewer: Examples! how do we actually use these? where do they take us/what does it look like when they do what they do?

> ### Syntax

Reviewer: Examples! how do we actually use these? where do they take us/what does it look like when they do what they do?

**Proceed with caution!** Using the `Sudo` command is decidedly not best data security practice, but might need to be invoked as you continue to learn bash, as permissioning can be quite granular and dynamic.
</div>

### Quiz: Permissions

> Which numerical sequence grants no restrictions on permissions and should generally be avoided?

Reviewer: The answer to this is in a link the learner has to click. Since it is important enough to have its own quiz question, let's put that information directly into the previous page.

> Which numerical sequence grants no restrictions on permissions and should generally be avoided?

Reviewer: The answer to this is in a link the learner has to click. Since it is important enough to have its own quiz question, let's put that information directly into the previous page.

> - If the module difficulty was appropriate

Reviewer: Grab the Feedback structure from the most recent sample module, I think it is a little different (links yaml, etc)

> - If the module difficulty was appropriate

Reviewer: Grab the Feedback structure from the most recent sample module, I think it is a little different (links yaml, etc)
